homeassistant/components/deconz/__init__.py

```diff
@@ -33,17 +36,28 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
     if not config_entry.options:
         await async_update_master_gateway(hass, config_entry)
 
-    gateway = DeconzGateway(hass, config_entry)
-    if not await gateway.async_setup():
-        return False
-
-    if not hass.data[DOMAIN]:
-        async_setup_services(hass)
+    try:
+        api = await get_deconz_session(hass, config_entry.data)
+    except CannotConnect as err:
+        raise ConfigEntryNotReady from err
+    except AuthenticationRequired as err:
+        raise ConfigEntryAuthFailed from err
+
+    gateway = hass.data[DOMAIN][config_entry.entry_id] = DeconzGateway(
+        hass, config_entry, api
+    )
 
-    hass.data[DOMAIN][config_entry.entry_id] = gateway
+    config_entry.add_update_listener(gateway.async_config_entry_updated)
+    hass.config_entries.async_setup_platforms(config_entry, PLATFORMS)
 
+    await async_setup_events(gateway)
     await gateway.async_update_device_registry()
 
+    if len(hass.data[DOMAIN]) == 1:
```

Comment:
This check is not good. Reloading the config entry will now setup the services again.

Reply:
That's not an issue. Services will be removed when the final entry is unloaded which happens also during reload

```
if not hass.data[DOMAIN]:
         async_unload_services(hass)
```